packages/common/async/src/task-scheduling.ts

```diff
@@ -141,3 +142,60 @@ export const scheduleExponentialBackoffTaskInterval = (
     clearTimeout(timeoutId);
   });
 };
+
+const IDLE_DETECTION_INTERVAL = 200;
+const IDLE_THRESHOLD = IDLE_DETECTION_INTERVAL * 2;
+
+// Unix timestamp of the last time the idle detection task was run.
+// NOTE: Date.now() is used instead of performance.now() because the latter is significantly slower.
+let lastIdleTime = Date.now();
+const tasksOnIdle: (() => void)[] = [];
+let idleTimeout: NodeJS.Timeout | null = null;
+
+// If the timer is delayed, we know that the thread is saturated.
+const scheduleIdleTask = () => {
+  idleTimeout = setTimeout(() => {
+    lastIdleTime = Date.now();
+
+    // Run the tasks that were scheduled to run on idle.
+    const tasksToRun = Array.from(tasksOnIdle);
+    tasksOnIdle.length = 0;
+    for (const task of tasksToRun) {
```

Comment:
let's also check whether we're not taking too long after each task and potentially yield and continue on the next idle period